README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# etrader

<!-- badges: start -->

![CRAN
Version](https://www.r-pkg.org/badges/version/etrader?color=green) ![Dev
Version](https://img.shields.io/badge/github-0.1.4-blue.svg)
![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/etrader)
<!-- badges: end -->

R package for the ETRADE API, facilitating authentication, trading,
pricing data, account balances, positions, order history, option chains,
and more. After creating an account, obtain [developer authorization
keys](https://developer.etrade.com/getting-started) to use the API. This
package was built for personal use tokens not vendor tokens. See this
[article](https://exploringfinance.github.io/posts/2020-11-21-the-ultimate-dollar-cost-averaging-strategy/)
for a full example of obtaining keys, logging in, and automating a trade
strategy in production.

## Disclosure

This software is in no way affiliated, endorsed, or approved by ETRADE
or any of its affiliates. It comes with absolutely no warranty and
should not be used in actual trading unless the user can read and
understand the source code. The functions within this package have been
tested under basic scenarios. There may be bugs or issues that could
prevent a user from executing trades or canceling trades. It is also
possible trades could be submitted in error. The user will use this
package at their own risk.

It is highly recommended to use the
[sandbox](https://developer.etrade.com/getting-started/developer-guides#tab_1)
environment prior to trading. Please note, as stated in the
documentation, the sandbox environment will return trades different from
what was entered. The sandbox is to verify the entry is valid.

## Installation

You can install etrader using:

``` r
# Available on CRAN
install.packages("etrader")

# Install development version - same as CRAN
# install.packages("devtools")
devtools::install_github("exploringfinance/etrader")

```

## Authentication

Accessing the ETRADE API requires the use of authorization keys that can
be obtained through the [ETRADE Developer
site](https://developer.etrade.com/getting-started). To use the API, the
user will need to manually log into their ETRADE account at least once a
day; however, this can be automated using `RSelenium` and `Docker` as
shown in this
[article](https://exploringfinance.github.io/posts/2020-11-21-the-ultimate-dollar-cost-averaging-strategy/).

Once authorization tokens have been obtained, authentication can be
achieved in three simple steps: setting credentials, generating a login
URL, and requesting access tokens. Below are all the steps required.
Jump to `Authentication Example` to see the R code example.

1.  Visit the [ETRADE Developer
    site](https://developer.etrade.com/getting-started)
2.  Complete the API User Intent Survey, API Agreement, and Annual
    Attestation
3.  This will generate a key and secret for both a sandbox and
    production environment
4.  Use `etrd_auth_credentials` to set the key and secret for the
    environment
5.  Generate a login URL using `etrd_auth_login_url`
6.  After logging in and accepting, pass the verification code to
    `etrd_auth_access_token`

This will generate Access Tokens that will be good until midnight
Eastern Time. Please be sure to indicate at each step whether sandbox or
production tokens are being used or the authentication will fail.
Production environment is assumed by default.

Steps 1-3 only need to be completed once. 4-6 need to be completed each
trading day but can be automated with `RSelenium` and `Docker`.

## Sandbox Authentication Example

Below is an example to log into the Sandbox environment. This is used to
test order entry and other calls that may be made. To log into
production, ensure to use production API keys and also remove “sandbox =
TRUE”. The entire authentication process must be completed within a few
minutes or the tokens will expire.

Once the authentication is complete, the access tokens are valid for the
rest of the trading day unless there is no activity for two hours.

``` r
# --------- Step 1 -----------
# Obtain a key and secret from ETRADE (https://developer.etrade.com/getting-started)
# Set the key and secret as credentials into an oauth app using etrd_auth_credentials
# sandbox flag defaults to production

# Retrieve ETRADE credentials from secure location
sandbox = TRUE
etkeys = readRDS('/home/rstudio/Secure/etrd_login_ab.rds')
etrade_cred = etrd_auth_credentials(etkeys$sand_key, etkeys$sand_secret, sandbox = sandbox)

# --------- Step 2 -----------
# Generate a URL and log into ETRADE (the site will open automatically)
# The URL may need to be re-pasted after login if you end up on a standaed ETRADE screen
# Credentials are passed by default

loginURL = etrd_auth_login_url(sandbox = sandbox)

# --------- Step 3 -----------
# After a successful login, paste the verification code as shown

access_tok = etrd_auth_access_token(verif_code = 'XX123',sandbox = sandbox)

# If successful, the rest of the etrader functions can now be used
# This process can be automated using Docker and Selenium

# --------- Optional Step 4 -----------
# Tokens expire at midnight Eastern each day and a manual login is required next day
# If two hours elapsed within the trading days, the tokens will be inactivated
# Reactive using etrd_auth_manage_token where tokens need to be explicitly passed
etrd_auth_manage_token(access_tokens = access_tok, etrade_cred = etrade_cred, sandbox = sandbox)
```

## Account Data

Get account data. Access tokens are passed by default.

``` r
# Auth parameters are set by default
sandbox = TRUE
actdata = etrd_account_list(sandbox = sandbox)

str(actdata)

# 'data.frame': 4 obs. of  9 variables:
# $ accountId      : chr  "xxxx"
# $ accountIdKey   : chr  "xxxx" 
# $ accountMode    : chr  "IRA" "BROKERAGE" "CASH" "CASH"
# $ accountDesc    : chr  "Brokerage" "Complete Savings" "INDIVIDUAL" "INDIVIDUAL"
# $ accountName    : chr  "NickName-1" "NickName-2" "NickName-3" ""
# $ accountType    : chr  "MARGIN" "INDIVIDUAL" "INDIVIDUAL" "CASH"
# $ institutionType: chr  "BROKERAGE" "BROKERAGE" "BROKERAGE" "BROKERAGE"
# $ accountStatus  : chr  "ACTIVE" "ACTIVE" "ACTIVE" "CLOSED"
# $ closedDate     : int  0 0 0 1

# Get account balances
account_id = actdata$accountIdKey[1]
etrd_account_balance(account_id, sandbox = sandbox)
```

## Trading

Trade functions have been split up into three categories. Equities,
Mutual Funds, and Options. The order entry for `etrader` has been
greatly simplified compared to what the [ETRADE API
offers](https://apisb.etrade.com/docs/api/order/api-order-v1.html#/definition/orderPreview).
Single order entry is the functionality currently built. Use
`etrd_cancel_order` to cancel open orders.

*Note: the sandbox environment will return previews and order outputs
different than what is submitted. Most submissions sent for execution
will fail. The objective is to confirm the order syntax is correct and
not validate entry. If an order preview is successfully generated then
the syntax works. The next recommended step is to place orders when the
market is closed in the production environment to ensure the order entry
comes through correctly.*

``` r
# Ensure trades are entered in sandbox environment
sandbox = TRUE
actlist = etrd_account_list(sandbox = sandbox)
account = actlist$accountIdKey[1]

# Specify ticker and preview type. Different ticker will be returned in sandbox
# Please note: I am not recommending SPLG I am only using it as an example
ticker = 'SPLG'
previewOrder = 'df'

#### ------- Equity orders ----------- ####
# Limit buy order
limiteq = etrd_place_eq_order(account = account,
                    symbol = ticker,
                    quantity = 5,
                    orderAction = 'buy',
                    priceType = 'limit',
                    limitPrice = 20,
                    previewOrder = previewOrder,
                    sandbox = sandbox)

# Cancel open limit order
etrd_cancel_order(limiteq$accountidKey, limiteq$orderId, sandbox = sandbox)

# stop sell order
etrd_place_eq_order(account = account,
                    symbol = ticker,
                    quantity = 5,
                    orderAction = 'sell',
                    priceType = 'stop',
                    stopPrice = 8,
                    previewOrder = previewOrder,
                    sandbox = sandbox)

# Trailing stop buy
etrd_place_eq_order(account = account,
                    symbol = ticker,
                    quantity = 5,
                    orderAction = 'buy',
                    priceType = 'TRAILING_STOP_PRCT',
                    stopPrice = 1, # Will be percent or dollar value based on priceType
                    previewOrder = previewOrder,
                    sandbox = sandbox)
                  
# market order with no preview - this will be sent immediately for execution
# PLEASE NOTE: THIS ORDER ENTRY IS NOT IN THE SANDBOX ENVIRONMENT. 
# IF THIS ORDER IS COPY/PASTED IT WILL BE EXECUTED IN PRODUCTION.
etrd_place_eq_order(account = account,
                    symbol = ticker,
                    quantity = 1,
                    orderAction = 'buy',
                    priceType = 'market',
                    previewOrder = 'none')                    


#### ------- Mutual Fund orders ----------- ####
# Please note: I am not recommending SWTSX, I am only using it as an example
# Mutual fund sell everything
mforder = etrd_place_mf_order(account = account,
                              symbol = 'SWTSX',
                              quantity = 1,
                              quantityType = 'ALL_I_OWN',
                              mfTransaction = 'sell',
                              previewOrder = 'df',
                              sandbox = sandbox)

# Mutual fund buy
mforder = etrd_place_mf_order(account = account,
                              symbol = 'SWTSX',
                              quantityType = 'DOLLAR',
                              quantity = 2,
                              mfTransaction = 'buy',
                              previewOrder = 'df',
                              sandbox = sandbox)
                              
#### ------- Option orders ----------- ####
# Please note: I am not recommending FB options, I am only using it as an example
# Buy to open
  etrd_place_optn_order(account = account,
                        symbol = 'FB',
                        callPut = 'call',
                        expiryYear = '2018',
                        expiryMonth = '12',
                        expiryDay  = '21',
                        strikePrice = '140',
                        quantity = 1,
                        orderAction = 'BUY_OPEN',
                        priceType = 'market',
                        previewOrder = 'df',
                        sandbox = sandbox)

# Sell to open
  etrd_place_optn_order(account = account,
                        symbol = 'FB',
                        callPut = 'call',
                        expiryYear = '2018',
                        expiryMonth = '12',
                        expiryDay  = '21',
                        strikePrice = '140',
                        quantity = 1,
                        orderAction = 'SELL_OPEN',
                        priceType = 'market',
                        previewOrder = 'df',
                        sandbox = sandbox)

```

## Orders and transactions

Search for orders and transactions. Get detailed transaction data for a
specific transaction.

``` r
actlist = etrd_account_list(sandbox = sandbox)

# Filter for executed 
filterOrd = etrd_order_history(actlist$accountIdKey[1], transactionType = 'BUY', 
                              securityType = 'mf', status = 'EXECUTED', 
                              fromDate = Sys.Date()-30, toDate = Sys.Date(), sandbox = sandbox)
  
  
# Get a list of transactions
transdf = etrd_transactions(actlist$accountIdKey[1], fromDate = '2020-01-01', sandbox = sandbox)

# Get transaction details
transDet_df = etrd_transaction_detail(actlist$accountIdKey[1], transdf$transactionId[1], sandbox = sandbox)
```

## Market Data

Get quotes (delayed or live depending on the user agreement with ETRADE)
and Option Chains.

``` r
  # Quotes for up to 50 securities
  qts = etrd_market_quote(c('AAPL','MSFT','SPY'), sandbox = sandbox)
 
  # Non filtered option chain
  chain_def = etrd_option_chain('SPY', sandbox = sandbox)
  
  # Filtered Option Chain
   chain_filt = etrd_option_chain(symbol = 'AAPL',
                                 expiryYear = 2021,
                                 expiryMonth = 6,
                                 noOfStrikes = 20,
                                 chainType = 'call',
                                 sandbox = sandbox)
```


NEWS.md
# etrader 0.1.5

## Release Notes and News

### etrader 0.1.5 01/03/2023

Improved object test using methods::is()

### etrader 0.1.4 12/15/2022

Fixed an issue related to authentication headers.


### etrader 0.1.3 02/21/2021

Change Maintainer and added example under
portfolio to convert list to a data frame

### etrader 0.1.2 11/25/2020

Fixes for CRAN. Released with commit f5045f0.

### etrader 0.1.1 11/23/2020

Bug fix for trailing stop orders.

### etrader 0.1.0 11/22/2020

Initial release includes basic functionality and API calls. Authentication,
Placing Trades, Canceling Trades, and pulling account details are all included
with the initial release.

Complex option order entry is not included.

Disclosure: 
This software is in no way affiliated, endorsed, or approved by ETRADE or any 
of its affiliates. It comes with absolutely no warranty and should not be used 
in actual trading unless the user can read and understand the source code. The 
functions within this package have been tested under basic scenarios. There may 
be bugs or issues that could prevent a user from executing trades or canceling 
trades. It is also possible trades could be submitted in error. The user will 
use this package at their own risk.


tests/testthat.R
library(testthat)
library(etrader)

test_check("etrader")


R/trading.R
#' Place an Equity Order
#'
#' Execute an equity order on the ETrade platform for the selected account.
#' Users are given the option to preview the order before submitting.
#'
#' @section Warning: TRADES THAT ARE SUCCESSFULLY ENTERED WILL BE SENT FOR
#'   EXECUTION. THIS FUNCTION HAS HUNDREDS OF POTENTIAL COMBINATIONS AND ONLY A
#'   HANDFUL HAVE BEEN TESTED. IT IS STRONGLY RECOMMENDED TO TEST THE DESIRED
#'   ORDER USING THE SANDBOX ENVIRONMENT. PLEASE NOTE THE SANDBOX ENVIRONMENT
#'   WILL NOT RETURN THE ORDER ENTERED, IT IS USED FOR CONFIRMING THE ORDER
#'   ENTRY HAS CORRECT SYNTAX. ORDERS CAN ALSO BE TESTED IN OFF MARKET HOURS ON
#'   A VERY SMALL QUANTITY WITH LITTLE MONEY AT STAKE. ANOTHER OPTION IS TO USE
#'   LIMIT ORDERS FAR FROM THE CURRENT PRICE. ETRADE HAS THEIR OWN ERROR
#'   HANDLING BUT IF A SUCCESSFUL COMBINATION IS ENTERED IT COULD BE EXECUTED
#'   UNINTENTIONALLY. DOUBLE CHECK ALL ENTRIES BEFORE SUBMITTING.
#'
#' @inheritParams etrd_account_balance
#' @param symbol The market symbol for the security being bought or sold
#' @param quantity The number of shares to buy or sell
#' @param orderAction The action that the broker is requested to perform: BUY,
#'   SELL, BUY_TO_COVER, SELL_SHORT
#' @param priceType The type of pricing: MARKET, LIMIT, STOP, STOP_LIMIT,
#'   TRAILING_STOP_CNST, TRAILING_STOP_PRCT, MARKET_ON_OPEN, MARKET_ON_CLOSE,
#'   LIMIT_ON_OPEN, LIMIT_ON_CLOSE
#' @param stopPrice The designated boundary price for a stop order. For trailing
#'   stop orders this will represent the dollar amount or percentage trailing
#'   value. Enter percentages as whole numbers.
#' @param limitPrice The highest price at which to buy or the lowest price at
#'   which to sell if specified in a limit order
#' @param stopLimitPrice 	The designated boundary price for a stop-limit order
#' @param quantityType The type of the quantity - QUANTITY, DOLLAR, ALL_I_OWN
#'   (note: DOLLAR is not yet an option)
#' @param orderTerm  The term for which the order is in effect:
#'   GOOD_UNTIL_CANCEL, GOOD_FOR_DAY, IMMEDIATE_OR_CANCEL, FILL_OR_KILL
#' @param marketSession The session in which the order will be placed: REGULAR,
#'   EXTENDED
#' @param allOrNone If TRUE, the transactions specified in the order must be
#'   executed all at once or not at all; default is FALSE
#' @param previewOrder Elect to preview the order before submitting.  Options
#'   are: 'df', 'json', or 'none' for an instant order entry. In a non
#'   interactive environment will default to 'none'. This also indicates how the
#'   order output is shown. json will return the output as a list, otherwise it
#'   will return as a dataframe
#'
#'
#' @return A list or data frame of the order entry details
#' @export
#'
#' @examples
#'
#' \dontrun{
#'
#' # Place Equity Limit order
#' acts = etrd_account_list()
#' account = acts$accountIdKey[1]
#' etrd_place_eq_order(account = account,
#'                     symbol = 'PSLV',
#'                     quantity = 1,
#'                     orderAction = 'buy',
#'                     priceType = 'limit',
#'                     limitPrice = 8,
#'                     previewOrder = 'df')
#'
#' }
etrd_place_eq_order = function(account,
                               symbol,
                               quantity,
                               orderAction,
                               priceType,
                               stopPrice = '',
                               limitPrice = '',
                               stopLimitPrice = '',
                               quantityType = 'quantity',
                               orderTerm = 'good_for_day',
                               marketSession= 'regular',
                               allOrNone= 'false',
                               previewOrder = 'df',
                               etrade_cred = NULL,
                               access_tokens = NULL,
                               sandbox = FALSE) {

    # Verify passed credentials or get credentials from options
    etrade_cred = hlp_etrd_check_cred(etrade_cred, sandbox)
    access_tokens = hlp_etrd_check_acctok(access_tokens, sandbox)

    # symbol = 'pslv'; quantity = 1; orderAction = 'buy'; priceType = 'market'; allOrNone= 'false'
    # quantityType = 'quantity'; orderTerm = 'good_for_day'; marketSession= 'regular'
    # stopPrice = ''; limitPrice = ''; stopLimitPrice = ''; sandbox = TRUE; previewOrder = 'df'
    clientOrdr = paste0('ordr',as.numeric(Sys.time()))

    # Collect order details
    orderList = list(
      PreviewOrderRequest = list(
        orderType = 'eq',
        clientOrderId = clientOrdr,
        Order = list(list(
          allOrNone = allOrNone,
          priceType = priceType,
          orderTerm = orderTerm,
          marketSession = marketSession,
          stopPrice = stopPrice,
          limitPrice = limitPrice,
          stopLimitPrice = stopLimitPrice,
          Instrument = list(list(
            Product = list(
              securityType = 'eq',
              symbol = symbol
            ),
            orderAction = orderAction,
            quantityType = quantityType,
            quantity = quantity
          ))
        ))
      )
    )

    hlp_etrd_place_order(account, stopPrice, orderList, clientOrdr, tolower(previewOrder), etrade_cred, access_tokens, sandbox)
}


#' Place a Mutual Fund Order
#'
#' Submit a mutual fund order on the ETrade platform for the selected account.
#' Users are given the option to preview the order before submitting. Mutual
#' fund orders must be received before 4pm or will be executed the following
#' day.
#'
#' @section Warning: TRADES THAT ARE SUCCESSFULLY ENTERED WILL BE SENT FOR
#'   EXECUTION. THIS FUNCTION HAS HUNDREDS OF POTENTIAL COMBINATIONS AND ONLY A
#'   HANDFUL HAVE BEEN TESTED. IT IS STRONGLY RECOMMENDED TO TEST THE DESIRED
#'   ORDER USING THE SANDBOX ENVIRONMENT. PLEASE NOTE THE SANDBOX ENVIRONMENT
#'   WILL NOT RETURN THE ORDER ENTERED, IT IS USED FOR CONFIRMING THE ORDER
#'   ENTRY HAS CORRECT SYNTAX. ORDERS CAN ALSO BE TESTED IN OFF MARKET HOURS ON
#'   A VERY SMALL QUANTITY WITH LITTLE MONEY AT STAKE. ETRADE HAS THEIR OWN
#'   ERROR HANDLING BUT IF A SUCCESSFUL COMBINATION IS ENTERED IT COULD BE
#'   EXECUTED UNINTENTIONALLY. DOUBLE CHECK ALL ENTRIES BEFORE SUBMITTING.
#'
#' @inheritParams etrd_place_eq_order
#' @param quantity The amount of the investment in either DOLLARS or
#'   SHARES depending on the input for quantityType
#' @param mfTransaction The transaction for the mutual fund order. Options:	BUY,
#'   SELL
#' @param reInvestOption Indicator flag to specify whether to reinvest profit on
#'   mutual funds. Options: REINVEST, DEPOSIT, CURRENT_HOLDING
#'
#' @return A list or data frame of the order entry details
#' @export
#'
#' @examples
#' \dontrun{
#'
#' # Place Mutual Fund Buy order for $10
#' acts = etrd_account_list()
#' account = acts$accountIdKey[1]
#' etrd_place_mf_order(account = account,
#'                     symbol = 'SWTSX',
#'                     quantityType = 'DOLLAR',
#'                     quantity = 10,
#'                     mfTransaction = 'buy',
#'                     reInvestOption = 'reinvest',
#'                     previewOrder = 'df')
#'
#' }
etrd_place_mf_order = function(account,
                               symbol,
                               quantity,
                               mfTransaction,
                               reInvestOption = 'REINVEST',
                               quantityType = 'DOLLAR',
                               previewOrder = 'df',
                               etrade_cred = NULL,
                               access_tokens = NULL,
                               sandbox = FALSE) {

  # Verify passed credentials or get credentials from options
  etrade_cred = hlp_etrd_check_cred(etrade_cred, sandbox)
  access_tokens = hlp_etrd_check_acctok(access_tokens, sandbox)

  # reInvestOption = 'REINVEST'; investmentAmount = '10'; symbol = 'SWTSX'
  # mfTransaction = 'BUY'; sandbox = TRUE; previewOrder = 'df'
  clientOrdr = paste0('ordr',as.numeric(Sys.time()))

  # Collect order details
  orderList = list(
    PreviewOrderRequest = list(
      orderType = 'MF',
      clientOrderId = clientOrdr,
      Order = list(list(
        reInvestOption = reInvestOption,
        investmentAmount = quantity,
        Instrument = list(list(
          Product = list(
            securityType = 'MF',
            symbol = symbol
          ),
          mfQuantity = list(cash = quantity),
          orderAction = mfTransaction,
          mfTransaction = mfTransaction,
          quantityType = quantityType
        ))
      ))
    )
  )

  # Mutual Funds do not require stop prices
  stopPrice = ''
  hlp_etrd_place_order(account, stopPrice, orderList, clientOrdr, tolower(previewOrder), etrade_cred, access_tokens, sandbox)
}


#' Place an Option Order
#'
#' Execute an option order on the ETrade platform for the selected account.
#' Users are given the option to preview the order before submitting. Note:
#' ETrade offers significantly more complex order structures than what is
#' offered in this function. See the
#' \href{https://apisb.etrade.com/docs/api/order/api-order-v1.html#}{ETrade
#' documentation} for more details on submitting complex option strategies.
#'
#' @section Warning: TRADES THAT ARE SUCCESSFULLY ENTERED WILL BE SENT FOR
#'   EXECUTION. THIS FUNCTION HAS HUNDREDS OF POTENTIAL COMBINATIONS AND ONLY A
#'   HANDFUL HAVE BEEN TESTED. IT IS STRONGLY RECOMMENDED TO TEST THE DESIRED
#'   ORDER USING THE SANDBOX ENVIRONMENT. PLEASE NOTE THE SANDBOX ENVIRONMENT
#'   WILL NOT RETURN THE ORDER ENTERED, IT IS USED FOR CONFIRMING THE ORDER
#'   ENTRY HAS CORRECT SYNTAX. ORDERS CAN ALSO BE TESTED IN OFF MARKET HOURS ON
#'   A VERY SMALL QUANTITY WITH LITTLE MONEY AT STAKE. ANOTHER OPTION IS TO USE
#'   LIMIT ORDERS FAR FROM THE CURRENT PRICE. ETRADE HAS THEIR OWN ERROR
#'   HANDLING BUT IF A SUCCESSFUL COMBINATION IS ENTERED IT COULD BE EXECUTED
#'   UNINTENTIONALLY. DOUBLE CHECK ALL ENTRIES BEFORE SUBMITTING.
#'
#' @inheritParams etrd_place_eq_order
#' @param orderAction The action that the broker is requested to perform:
#'   BUY_OPEN, BUY_CLOSE, SELL_OPEN, SELL_CLOSE
#' @param callPut The option type:	CALL, PUT
#' @param expiryYear The four-digit year the option will expire
#' @param expiryMonth The month (1-12) the option will expire
#' @param expiryDay The day (1-31) the option will expire
#' @param strikePrice The strike price for the option
#'
#' @return A list or data frame of the order entry details
#' @export
#'
#' @examples
#' \dontrun{
#'
#' # Place Option Buy_to_open order
#' acts = etrd_account_list()
#' account = acts$accountIdKey[1]
#' etrd_place_optn_order(account = account,
#'                       symbol = 'FB',
#'                       callPut = 'call',
#'                       expiryYear = '2018',
#'                       expiryMonth = '12',
#'                       expiryDay  = '21',
#'                       strikePrice = '140',
#'                       quantity = 1,
#'                       orderAction = 'BUY_OPEN',
#'                       priceType = 'market',
#'                       previewOrder = 'df')
#'
#' }
etrd_place_optn_order = function(account,
                                 symbol,
                                 callPut,
                                 expiryYear,
                                 expiryMonth,
                                 expiryDay,
                                 strikePrice,
                                 quantity,
                                 orderAction,
                                 priceType,
                                 stopPrice = '',
                                 limitPrice = '',
                                 stopLimitPrice = '',
                                 quantityType = 'quantity',
                                 orderTerm = 'good_for_day',
                                 marketSession= 'regular',
                                 allOrNone= 'false',
                                 previewOrder = 'df',
                                 etrade_cred = NULL,
                                 access_tokens = NULL,
                                 sandbox = FALSE) {

  # Verify passed credentials or get credentials from options
  etrade_cred = hlp_etrd_check_cred(etrade_cred, sandbox)
  access_tokens = hlp_etrd_check_acctok(access_tokens, sandbox)

  # symbol = 'pslv'; quantity = 1; orderAction = 'buy_open'; priceType = 'market'; allOrNone= 'false'
  # quantityType = 'quantity'; orderTerm = 'good_for_day'; marketSession= 'regular'
  # stopPrice = ''; limitPrice = ''; stopLimitPrice = ''; sandbox = TRUE; previewOrder = 'df'
  # callPut = 'CALL'; expiryYear = 2018; expiryMonth = 12; expiryDay = 21; strikePrice = 140

  clientOrdr = paste0('ordr',as.numeric(Sys.time()))

  # Collect order details
  orderList = list(
    PreviewOrderRequest = list(
      orderType = 'optn',
      clientOrderId = clientOrdr,
      Order = list(list(
        allOrNone = allOrNone,
        priceType = priceType,
        orderTerm = orderTerm,
        marketSession = marketSession,
        stopPrice = stopPrice,
        limitPrice = limitPrice,
        stopLimitPrice = stopLimitPrice,
        Instrument = list(list(
          Product = list(
            securityType = 'optn',
            symbol = symbol,
            callPut = callPut,
            expiryYear = expiryYear,
            expiryMonth = expiryMonth,
            expiryDay = expiryDay,
            strikePrice = strikePrice
          ),
          orderAction = orderAction,
          orderedQuantity = quantity,
          quantity = quantity
        ))
      ))
    )
  )



  hlp_etrd_place_order(account, stopPrice, orderList, clientOrdr, tolower(previewOrder), etrade_cred, access_tokens, sandbox)
}

#' Cancel an existing order
#'
#' Cancel an open order that has been submitted. Note: Verify the cancel request
#' was received and processed
#'
#' @inheritParams etrd_account_balance
#' @param orderId Order confirmation Id for the order placed.
#'
#' @return a response validating that the order has been canceled
#' @export
#'
#' @examples
#' \dontrun{
#' # Place Mutual Fund Buy order for $10
#' acts = etrd_account_list()
#' account = acts$accountIdKey[1]
#' mforder = etrd_place_mf_order(account = account,
#'                     symbol = 'SWTSX',
#'                     quantityType = 'DOLLAR',
#'                     investmentAmount = 10,
#'                     mfTransaction = 'buy',
#'                     reInvestOption = 'reinvest',
#'                     previewOrder = 'df')
#'
#' etrd_cancel_order(mforder$accountidKey, mforder$orderid.orderId)
#'
#' }
etrd_cancel_order = function(account,
                             orderId,
                             output = c('df','list'),
                             access_tokens = NULL,
                             etrade_cred = NULL,
                             sandbox = FALSE) {

  # Verify passed credentials or get credentials from options
  etrade_cred = hlp_etrd_check_cred(etrade_cred, sandbox)
  access_tokens = hlp_etrd_check_acctok(access_tokens, sandbox)

  # Generate signature
  sb = ifelse(sandbox,'sb','')
  cancelURL = paste0('https://api',sb,'.etrade.com/v1/accounts/',account,'/orders/cancel')
  signature = httr::oauth_signature(
    url = cancelURL,
    method = 'PUT',
    app = etrade_cred,
    token = access_tokens$oauth_token,
    token_secret = access_tokens$oauth_token_secret
  )

  # pull account list
  cancel_req = httr::PUT(cancelURL,
                         hlp_etrd_auth_headers(signature),
                         body=list(CancelOrderRequest = list(orderId = orderId)),encode='json')

  # Check status
  hlp_etrd_status(cancel_req)
  cancel_req_resp = httr::content(cancel_req)

  # Return output
  cancel_req_resp

}





# ---------- Helper Trade Function 1 hlp_etrd_trd_status ----------------
# Check if trade function did not return 200 send back error code
hlp_etrd_trd_status = function(x,msg=NULL){

  # Check if status code is 200 or 201 (201 is specific to orders)
  SC = x$status_code
  if (SC!=200) {

    # Default to TD Error message and append custom if needed
    ErrMsg = httr::content(x)
    stop(paste0(SC,' - ',ErrMsg), call. = FALSE)
  }

}


# ---------- Helper Trade Function 2 hlp_etrd_place_order ----------------
# Place order collected from trade function
hlp_etrd_place_order = function(account,
                                stopPrice, # For trailing stop this must be explicitly passed
                                orderList,
                                clientOrdr,
                                previewOrder,
                                etrade_cred,
                                access_tokens,
                                sandbox) {

    # Create URL and generate signature
    sb = ifelse(sandbox,'sb','')
    ordrprevURL = paste0('https://api',sb,'.etrade.com/v1/accounts/',account,'/orders/preview.json')
    sign = httr::oauth_signature(
      url = ordrprevURL,
      method = 'POST',
      app = etrade_cred,
      token = access_tokens$oauth_token,
      token_secret = access_tokens$oauth_token_secret
    )

    # Post order for preview
    order_prev = httr::POST(ordrprevURL,hlp_etrd_auth_headers(sign),body=orderList,encode='json')

    # Check preview status and pull content
    hlp_etrd_trd_status(order_prev)
    ordrCon = httr::content(order_prev)


    # If interactive and preview requested, then show order and ask for confirmation
    if (previewOrder != 'none' & interactive()==TRUE) {
      # Display order details
      dfOut = data.frame(orderDetails = c(accountid = ordrCon$PreviewOrderResponse$accountId,
                                          unlist(ordrCon$PreviewOrderResponse$Order)))
      jsonOut = jsonlite::toJSON(ordrCon, pretty = TRUE, auto_unbox = TRUE)

      printOut = if(previewOrder == 'json') {jsonOut} else {dfOut}
      print(printOut)

      # Show order output as data frame
      # purrr::map_df(ordrCon, ~as.data.frame(.))

      # Ask for confirmation
      prevorder <- utils::menu(c('Yes','No'),
                           title = paste0('\nDo you want to place the order as shown above?'))

      # If user selects No or exits, do not place order

      ####****** Revisit with cancel order and indicate order is canceled
      if (prevorder %in% c(0,2)) stop('Order Not Placed', call. = FALSE)
    }


    # Ensure stop price matches original entry - necessary for trailing stops
    ordrCon$PreviewOrderResponse$Order[[1]]$stopPrice <- stopPrice

    # Create list of order to place using output
    placeList <- list(
      PlaceOrderRequest = list(
        orderType = ordrCon$PreviewOrderResponse$orderType,
        clientOrderId = clientOrdr,
        PreviewIds = ordrCon$PreviewOrderResponse$PreviewIds,
        Order = ordrCon$PreviewOrderResponse$Order
      )
    )

    # Generate URL and signature for placing order
    sb <- ifelse(sandbox,'sb','')
    ordrplaceURL <- paste0('https://api',sb,'.etrade.com/v1/accounts/',account,'/orders/place.json')
    sign <- httr::oauth_signature(
      url = ordrplaceURL,
      method = 'POST',
      app = etrade_cred,
      token = access_tokens$oauth_token,
      token_secret = access_tokens$oauth_token_secret
    )

    # Place Order
    order_place <- httr::POST(ordrplaceURL,hlp_etrd_auth_headers(sign),body=placeList,encode='json')

    # Verify order placed successfully
    hlp_etrd_trd_status(order_place)

    # Return order content
    orderList <- httr::content(order_place)

    orderDF <- data.frame(accountid = orderList$PlaceOrderResponse$accountId,
                       accountidKey = account,
                       orderid = orderList$PlaceOrderResponse$OrderIds[[1]],
                       t(unlist(orderList$PlaceOrderResponse$Order)))

    printOut <- if(previewOrder == 'json') {orderList} else {orderDF}
    printOut


}
